### ISSUE TITLE
Code generation for sequences fails for SQL Server 2014

### ISSUE THREAD

**Issue by PaulRambags** [jOOQ/jOOQ#14425]
In jOOQ 3.17.3, 3.17.6 and 3.18.0-SNAPSHOT-20221217003226, the following exception occurs when trying to generate code for Sequences in a SQL Server 2014 database:

`SQL exception            : Exception while executing meta query: Arithmetic overflow error converting float to data type numeric.`

This is the problematic code that was reported by the jOOQ code generation:

```sql
select
  db_name() catalog,
  alias_92986274.name,
  sys.sequences.name,
  alias_31823739.name,
  alias_31823739.precision,
  alias_31823739.scale,
  nullif(
    cast(sys.sequences.start_value as bigint),
    1
  ) start_value,
  nullif(
    cast(sys.sequences.increment as bigint),
    1
  ) increment,
  nullif(
    cast(sys.sequences.minimum_value as bigint),
    -power(
      cast(2 as numeric),
      ((alias_31823739.max_length * 8) - 1)
    )
  ) minimum_value,
  nullif(
    cast(sys.sequences.maximum_value as numeric(38)),
    (power(
      cast(2 as numeric),
      ((alias_31823739.max_length * 8) - 1)
    ) - 1)
  ) maximum_value,
  sys.sequences.is_cycling,
  sys.sequences.cache_size
from (
  sys.sequences
    join sys.schemas alias_92986274
      on sys.sequences.schema_id = alias_92986274.schema_id
    join sys.types alias_31823739
      on sys.sequences.user_type_id = alias_31823739.user_type_id
  )
where alias_92986274.name in (N'dbo')
order by alias_92986274.name, sys.sequences.name
```

When I run the above query in a SQL Server 2014 database, I get an "Arithmetic overflow error converting float to data type numeric." The query runs fine in a SQL Server 2019 database (see below). Remark: the sequences have different values in that database - so this might not have to do with a difference between the 2014 and 2019 versions of SQL Server - however, the sequence values are larger in the 2019 server (see  below).

The sequence generation worked fine under jOOQ 3.11.9

More details:

This query gives details about all existing sequences:

```sql
SELECT
  name,
  cast(start_value AS NUMERIC)   AS start_value,
  cast(increment AS NUMERIC)     AS increment,
  cast(current_value AS NUMERIC) AS current_value
FROM sys.sequences;
```

This is the result on the SQL Server 2014:
![image](https://user-images.githubusercontent.com/29857059/208616949-3ccabdda-9202-4529-8308-94d4caadf80b.png)

This is the result on the SQL Server 2019:
![image](https://user-images.githubusercontent.com/29857059/208617209-116086fd-2fae-4d6a-a442-8c49a5d8df06.png)

This is the result of the problematic query when run in the SQL Server 2019:
![image](https://user-images.githubusercontent.com/29857059/208617658-bcf547b3-a072-4ff4-8bb5-0486ff1fbdb9.png)

**Comment by lukaseder** [comment]
Thanks a lot for your report. I will look into this right away.

**Comment by lukaseder** [comment]
I can reproduce this simply, with https://dbfiddle.uk/44LwxUcZ. It doesn't happen with any more recent SQL Server version.

I'm not sure we really need the `CAST`. Using `coerce()` should probably suffice to make sure we get the right data type.

**Comment by lukaseder** [comment]
A simpler query that reproduces the problem in SQL Server 2014 is this one:

```sql
select power(cast(2 as numeric(38)), cast(t.max_length as numeric(38)) * 8 - 1)
  from sys.types t
  join sys.sequences s on s.user_type_id = t.user_type_id;
```

This seems to be a bug in my opinion, given that this works in later versions of SQL Server? Some strange behaviour when using `CAST()` on the data type of the `SYS.TYPES.MAX_LENGTH` column. I'll check if any workaround is possible.

**Comment by lukaseder** [comment]
This would be a workaround:

```sql
declare @t table (a numeric, b numeric);

insert into @t
select cast (2 as numeric),  cast(t.max_length * 8 - 1 as numeric)
 from sys.types t
  join sys.sequences s on s.user_type_id = t.user_type_id;

select power(a, b) from @t;
```

**Comment by lukaseder** [comment]
Or, a `CASE` expression:

```sql
select 
  case t.max_length
    when 1 then 127
    when 2 then 32768
    when ...
  end
from sys.types t
join sys.sequences s on s.user_type_id = t.user_type_id;
```

**Comment by lukaseder** [comment]
OK, this query seems to work!

```sql
select
  db_name() catalog,
  alias_92986274.name,
  sys.sequences.name,
  alias_31823739.name,
  alias_31823739.precision,
  alias_31823739.scale,
  nullif(
    cast(sys.sequences.start_value as bigint),
    1
  ) start_value,
  nullif(
    cast(sys.sequences.increment as bigint),
    1
  ) increment,
  nullif(
    cast(sys.sequences.minimum_value as varchar(max)),
    case alias_31823739.max_length
      when 1 then '0'
      when 2 then '-32768'
      when 4 then '-2147483648'
      when 8 then '-9223372036854775808'
      else ('-' + replicate('9', sys.sequences.precision))
    end
  ) minimum_value,
  nullif(
    cast(sys.sequences.maximum_value as varchar(max)),
    case alias_31823739.max_length
      when 1 then '255'
      when 2 then '32767'
      when 4 then '2147483647'
      when 8 then '9223372036854775807'
      else replicate('9', sys.sequences.precision)
    end
  ) maximum_value,
  sys.sequences.is_cycling,
  sys.sequences.cache_size
from (
  sys.sequences
    join sys.schemas alias_92986274
      on sys.sequences.schema_id = alias_92986274.schema_id
    join sys.types alias_31823739
      on sys.sequences.user_type_id = alias_31823739.user_type_id
  )
where alias_92986274.name in (N'dbo')
order by alias_92986274.name, sys.sequences.name
```

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.7 (https://github.com/jOOQ/jOOQ/issues/14427), and 3.16.13 (https://github.com/jOOQ/jOOQ/issues/14428)
